README.md
## Description

IOport Blip Incremental is a frequent iterative pre-release repo that can be installed by changing the ioport-hub _Dist/install.js_ file.

Change code section as shown:


* Comment out: //  repoUrl: "https://github.com/Balguardth/ioport-blip.git"
* Add: repoUrl: "https://github.com/Balguardth/ioport-blip-incremental.git"


## Installation

Follow the [documentation](https://secure.ioport.com/Documentation) for setup instructions and more information.  Note: The local version of the documentation will usually only reflect the latest official release from the _ioport-blip_ repo.




History.md
1.0.1 / 2020-07-23
===================

  * Initial release of the software.

1.0.2 / 2020-07-23
===================

  * Reduced archive size.
  * Updated server package dependencies.
  * Updated Help page with online links.

1.0.3 / 2020-07-28
===================
  * Updated BlipDynamoTabs menu.
    - Fixed jitter bug.
    - Added "linkAlternative" option to activate tabs.      
      - Solution allows for an alternative link to be added while keeping hostname constraint.
      - Updated "Documentation > Client-Side > Navigation Menu" with addition.

1.0.4 / 2020-07-31
===================
  * Updated BlipDynamoTabs menu.
    - Added "linkAlternative" array option to activate tabs.      
      - Can be used with a single URL entry or an array of URLs.
    - Added "urlParamParsers" to allow URL split tab activation.
  * Updated "Documentation".

1.0.5 / 2020-08-03
===================
  * Updated TemplateEngine.addStrReplace(id, obj).
    - Location: .\Framework\Server\Core\Common\Common_Template-Assembler.js.
    - Will now update existing entries.
      - The fragment cache flag needs to be false for live updates.
    - Can take either a single entry object or an array of objects when updating. Initial entry must still be an array.

1.0.6 / 2020-08-06
===================
  * Added try...catch to the framework require includes in File-Server.js.
    - Captures error events during compile time and is now part of the debug control system.
  * Added port variables to the GLOBAL object in ./Assets/Common/Js/Main.js.
  * Updated "Documentation".
    - Reflecting additions above.
    - Added new entry for Client-Side > AJAX > setPostResFunc(func).
      - Text pertaining to this.responseData in the comm object.

1.0.7 / 2020-08-18
===================
  * Updated code comments to reflect documentation.

1.0.8 / 2020-11-11
===================
  * Added JavaScript CSS style.display = 'none' assignment to all child nodes added to the shadow popup container (GLOBAL.popups.shadow.panel) when using GLOBAL.popups.shadow.hide().
  * Added GLOBAL.popups.shadow.flagContainerShadowClickIgnore to GLOBAL.popups.shadow object.
  * Added domain restriction to configured specification.
  * Added compensation code for various mobile browser quirks.
  * Added server notification page messages for:
    - Status code 500, server error.
    - Status code 404, page not found.
  * Documentation updates.
    - New code additions.
    - Corrected description for "static sendText(res, text)".

2.1.0 / 2023-02-07
===================
  * Refactored the codebase to run with the Yarn package manager as a workspace.
  * Refactored the IOport Blip modules implementation.  Know using "site docks" to avoid confusion with node_modules.
    - This is revamped for the most part.
  * Updated BlipDynamoTabs to display as expected when using HD video monitors.
  * Updated node-modules dependencies.
  * Added more host configuration options and URL handling.
  * Updated documentation.

2.5.1 / 2023-02-13
===================
  * Added configuration options to edit the log format.
  * Updated documentation.
  * Tweaked the Blip client-side theme.
  * Fixed bug in the URL redirect method.

2.5.2 / 2023-02-13
===================
  * Set the default logger time to military.

2.5.3 / 2023-02-22
===================
  * Added logger addition to page error handler.  URLs will now be included.
  * Added error handling to req header filtering.
  * Refactored require _.init(..)_ method relating to _Core_ and _SiteDock_ files.
    - Note: Functionality remains unchanged.

2.5.4 / 2023-02-27
===================
  * Refactored the init(...) process for _Core_ and _SiteDock_ scripts.
    - Note: Code integrity is maintained.
  * Added error handling to the _SiteDocks_ loading method that catches repeat installation attempts of the same site dock.
    - Renaming the site dock directory that the package was installed into will prevent this error notification.
  * Added blip.svar.flagSiteDockInstallOnStart and blip.svar.flagSiteDockLoadHanger71OnStart.
    - Enable/disable:
      - Installation of new packages on startup.
      - Loading of Hanger71 site docks on startup.
  * Added Hanger71 site dock package integrity.  All Hanger71 related packages are registered and managed within the hub application space.
    - Note: This helps ensure that the Hanger71 space is used how it is intended.

2.5.5 / 2023-03-xx (pre-release)
===================
  * Refactored the application connection verification and error handling method.
    - This can now be directly addressed in the site dock configuration file.
  * Added _thirdParty_ site dock integrity check.
    - The _SiteDock-Frame-Basic.zip.zbm_ reflects this.
  * Changed the identifier file token length to 16 instead of 64.
    - Delete the existing identifier file in the _./dock_ directory to generate a new one on server start-up.
    - The _blipSiteDock.svar.token_ is initialized with the identifier token.
  * Added callbacks when loading the different site dock types.
  * Refactored the registry cleaning method to wait until after all of the site docks are loaded.
  * Added _getIdents(...)_ to common utilities.
  * Added _getPackagesLoadStatus(...)_ to the template assembler.
    - This provides a way to check the loaded package conditions.
  * Added identifier token generation parameters to _blip.svar.identTokenParams_.
  * The object properties _blip.svar.packageName_ and _blipSiteDock.svar.packageName_ now get initialized from their respected criteria file.
  * Added _blip.utilites.renewIdentifierToken(...)_
    - Replaces the identifier token with a new one for:
      - Hub applications
      - Blip site docks
  * **UNDER DEVELOPMENT**
    - Blip site dock deployment station.
      * Allows a supply beacon to be downloaded that manages installations using split archive files.
      * Accessible through the blip site dock interface.


Server/File-Server.js
/************
 * IMPORTANT:  It is recommended that you read the documentation if you are thinking about editing this file.
 ************/
"user strict";

function initBlip() {

    const date = new Date();

    const ref = Object.freeze({
        loggerOutputTypeOptions: {
            none:           0,
            console:        1,
            file:           2,
            process:        3,
            consoleAndFile: 4,
            processAndFile: 5
        }
    });

    const svar = {
        date:                              date,
        appName:                           'IOport Blip',
        appPackageName:                    '',      // Assigned from the criteria file.
        version:                           '2.5.4', // {sip var="blipVersion" /}
        copyrightYear:                     ((date.getUTCFullYear() != '2020') ? '2020 - ' + date.getUTCFullYear(): '2020'),
        loggerOutputType:                  ref.loggerOutputTypeOptions.consoleAndFile,
        flagVerbose:                       true,    // Enable/Disable detailed running log
        flagCache:                         false,   // Enable/Disable all caching
        flagSiteDockInstallOnStart:        true,    // Enable/Disable installing new site dock packages when the server starts.
        flagSiteDockLoadHanger71OnStart:   true,    // Enable/Disable loading Hanger71 site docks when the server starts.
        testDelay:                         0,       // For simulating server request loads.
        favExt:                            '.png',
        cssExt:                            '.css',
        jsExt:                             '.js',
        registrationFileNameExt:           Object.freeze('.reg.json'),
        siteDocksPackageFileExts:          Object.freeze({zbm: '.zip.zbm', tbm: '.zip.tbm' }),
        siteDockTypes:                     Object.freeze({hanger71: 0, thirdParty: 1}),
        package:                           {type:{
                                                hub:              0,
                                                app:              1,
                                                siteDock:         2,
                                                siteDockHanger71: 3
                                              }
                                           },
        siteDocksPackageExtractCheckDelay: 1000,    // Check every minute
        siteDocksPackageExtractTimeout:    600000,  // Try for 10 minutes
        identTokenParams:                  {iterations: 120, tokenLength: 16, digestAlgorithm: 'sha256'},
        siteDocksBlipListen:               '127.0.0.1',
        siteDocksBlipDomainName:           'localhost',
        logDir:                            '/../../../Logs',
        logInfoFileNamePrefix:             'Log_Server-Info',
        logErrorFileNamePrefix:            'Log_Server-Error',
        logFileNameExt:                    '_%DATE%.log',
        logTimestamp:                      'YYYY-MM-DD HH:mm:ss.SS',
        logDatePattern:                    null,
        logFrequency:                      null,
        logFormat:                         ({ level, message, timestamp }) => {return `${timestamp} [${level}] ${message}`;},
        logMaxSize:                        '10m',
        logMaxFiles:                       '90d',
        logOutputMessages:                 {updatingConfigUsingExtension: 'Updating configuration using extension: $<fileName>',
                                            loadingSiteDockRequiredFile: 'Loading site dock: $<fileName>',
                                            loadingSiteDockFramework: 'Loading site dock framework: $<fileName>',
                                            loadingGenericRequiredFilesErr: 'Loading required files.'},
        pageNotification:                  {msg404: "Status code 404. This page can not be found.",
                                            msg500: "Status code 500. A server error has occurred."},
        templateLog:                       {cached: 'cached frag results: ', raw: 'raw frag results: '},
        sipTag:                            {open: '{sip var="', close: '" /}'},
        sipLoopTag:                        {beginOpen: '{sipLoop', beginClose: '}', endOpen: '{/sipLoop', endClose: '}'},
        sipLoopSplitTag:                   '{sipLoopSplit /}',
        sipLoopSplitTagVars:               {open: '{sipLoopSplit', close: '/}'},
        sssplitTag:                        '{sipSplit /}',
        siteDockInstanceNames:             [],
        siteDockInstancesLoading:          {hanger71: 0, thirdParty: 0},
        publicSalt:                        null
    };

    const path = {
        appDir:                                __dirname + '/../',
        hubDir:                                __dirname + '/../../',
        distDir:                               __dirname + '/../../Dist/',
        dotRegDir:                             __dirname + '/../../Dist/.reg/',
        dotAppDir:                             __dirname + '/../.app/',
        dotDockDir:                            '/.dock/',
        criteriaFileName:                      'criteria.json',
        identifierFileName:                    'identifier.json',
        nodePackageFileName:                   'package.json',
        modules:                               __dirname + '/node_modules/',
        serverClientDir:                       __dirname + '/../Client/SiteDocks/',
        coreDir:                               __dirname + '/Core/',
        coreCommonDir:                         __dirname + '/Core/Common/',
        configurationExtFileName:              __dirname + '/Configuration-Extension',
        utilitiesFileName:                     __dirname + '/Core/Common/Common_Utilities',
        templateAssemblerFileName:             __dirname + '/Core/Common/Common_Template-Assembler',
        siteDocksDir:                          __dirname + '/SiteDocks/',
        siteDocksDeployDir:                    __dirname + '/SiteDocks/Deploy/',
        siteDockHanger71Dir:                   __dirname + '/SiteDocks/Hanger71/',
        siteDockHanger71DeployDir:             __dirname + '/SiteDocks/Hanger71/Deploy/',
        siteDocksInstalledPackagesDir:         __dirname + '/SiteDocks/InstalledPackages/',
        siteDockHanger71InstalledPackagesDir:  __dirname + '/SiteDocks/Hanger71/InstalledPackages/',
        siteDocksFramesDir:                    __dirname + '/SiteDocks/ExamplePackageFrames/',
        siteDocksConfigFileName:               'SiteDock-Configuration',
        siteDocksFileName:                     __dirname + '/SiteDocks/SiteDocks'
    };

    const server = {
        winston:              Object.freeze(require('winston')),
        winstonRotateFile:    Object.freeze(require('winston-daily-rotate-file')),
        os:                   Object.freeze(require('os')),
        v8:                   Object.freeze(require('v8')),
        fs:                   Object.freeze(require('fs')),
        path:                 Object.freeze(require('path')),
        https:                Object.freeze(require('https')),
        http:                 Object.freeze(require('http')),
        util:                 Object.freeze(require('util')),
        inspect:              Object.freeze(require('util').inspect),
        express:              Object.freeze(require('express')),
        staticFile:           Object.freeze(require('connect-static-file')),
        moment:               Object.freeze(require('moment')),
        sizeOf:               Object.freeze(require('sizeof').sizeof),
        unzip:                Object.freeze(require('extract-zip')),
        yauzl:                Object.freeze(require('yauzl')),
        crypto:               Object.freeze(require('crypto')),
        logger:               null,
        loggerInfo:           null,
        loggerErr:            null,
        loggerInfoErr:        null,
        appConnQue:           null,           // Use for global tmp data.
        appConnQueReqLogging: {tmp: 'tmp'},   // Holds http and https connection params req and res
        loadFramework:        null
    };

    return { ref, svar, path, server };

}

let blip = initBlip();

 /************
 * Utilities
 ************/
loadUtilities();

async function loadUtilities() {

  blip['utilities'] = await require(blip.path.utilitiesFileName).init(blip);

  blip.utilities.integrityCheck(function (error) {

    if (error) return;
    loadConfigurationExt();

  });

}

 /************
  * Configuration Extension
  ************/
 function loadConfigurationExt() {    

    if(blip.server.fs.existsSync(blip.path.configurationExtFileName + '.js')) {

        let updateConfiguration = require(blip.path.configurationExtFileName);

        updateConfiguration.go(blip, function(error) {

            if(error) {

                blip.server.loggerErr(blip.svar.logOutputMessages.updatingConfigUsingExtension.replace(/\$<fileName>/,
                blip.path.configurationExtFileName));

            }

            loadRequiredFiles();

        });

    } else {

        loadRequiredFiles();

    }
}

function loadRequiredFiles() {

    // Freeze parameters
    blip.svar = Object.freeze(blip.svar);
    blip.path = Object.freeze(blip.path);
    blip.utilities = Object.freeze( blip.utilities);

    try{
        /************
         * Blip Site Dock File Require Initialization
         ************/
        blip.server['siteDocks'] = Object.freeze(require(blip.path.siteDocksFileName).init(blip));

        /************
         * Common Files
         ************/
        blip['templateAssembler'] = require(blip.path.templateAssemblerFileName).init(blip);

        if(blip.svar.flagSiteDockInstallOnStart) {

            blip.server.siteDocks.checkForSiteDocks(blip.svar.siteDockTypes.thirdParty, false, function(error) {

                if(error) return logCheckSiteDocksError(blip.svar.siteDockTypes.thirdParty);

                if(!blip.svar.flagSiteDockLoadHanger71OnStart) return;

                blip.server.siteDocks.checkForSiteDocks(blip.svar.siteDockTypes.hanger71, false, function(error) {
                    if(error) return logCheckSiteDocksError(blip.svar.siteDockTypes.hanger71);                    
                });

            });

        } else {

            blip.server.siteDocks.checkForSiteDocks(blip.svar.siteDockTypes.thirdParty, true, function(error) {

                if(error) return logCheckSiteDocksError(blip.svar.siteDockTypes.thirdParty);

                if(!blip.svar.flagSiteDockLoadHanger71OnStart) return;

                blip.server.siteDocks.checkForSiteDocks(blip.svar.siteDockTypes.hanger71, true, function(error) {
                    if(error) return logCheckSiteDocksError(blip.svar.siteDockTypes.hanger71);                    
                });

            });

        }

    } catch (error) {

        blip.server.loggerErr(blip.svar.logOutputMessages.loadingGenericRequiredFilesErr + ' ' + error);

    }

    function logCheckSiteDocksError(type) {

        blip.server.loggerErr('There was a problem when checking for ' + Object.keys(blip.svar.siteDockTypes)[type] + ' site dock.');

    }

}
